### PR TITLE
Implement Add and AddBatch for raw results

### PR DESCRIPTION
**Pull request by xichen2020**
cc @notbdu @black-adder 

This PR implements `Add` and `AddBatch` for raw results.

### CODE REVIEW

File: query/raw_result.go [hunk 1]
```diff
@@ -212,19 +213,98 @@ func (r *RawResults) IsOrdered() bool {
...
 // AddBatch adds a batch of raw results to the collection.
-// TODO(xichen): Implement this.
+// For unordered raw results, the incoming batch is unsorted:
+// - If the results have not reached limit yet, the incoming results are appended at the end
+//   until the limit is reached, after which the incoming results are dropped.
+// - Otherwise, the incoming results are dropped.
+// For ordered raw results, the incoming batch is sorted:
+// - If the results have not reached limit yet, the incoming results are added in order
+//   until the limit is reached, after which the incoming results are inserted and
+//   the results beyong limit are dropped.
```
Comment (black-adder) [inline query/raw_result.go]: s/beyong/beyond

Comment (xichen2020) [reply query/raw_result.go]: Good catch, thanks

File: query/raw_result.go [hunk 2]
```diff
@@ -212,19 +213,98 @@ func (r *RawResults) IsOrdered() bool {
...
+	// implementation for now. When we do, a templatized linked list might be a better choice
+	// for the result collection.
 	panic("not implemented")
 }
 
 // AddBatch adds a batch of raw results to the collection.
-// TODO(xichen): Implement this.
+// For unordered raw results, the incoming batch is unsorted:
+// - If the results have not reached limit yet, the incoming results are appended at the end
+//   until the limit is reached, after which the incoming results are dropped.
```
Comment (notbdu) [inline query/raw_result.go]: Haven't looked but should we add a `TODO` to add short circuiting logic to stop queries from executing once the result limit is reached?

File: query/raw_result.go [hunk 3]
```diff
@@ -212,19 +213,98 @@ func (r *RawResults) IsOrdered() bool {
...
+		existingIdx++
+	}
+	for newIdx < len(rr) && newResultsIdx < newSize {
+		r.cache[newResultsIdx] = rr[newIdx]
+		newResultsIdx++
+		newIdx++
+	}
+
+	// Swap data array and cache array to prepare for next add.
+	tmp := r.Data
```
Comment (notbdu) [inline query/raw_result.go]: `r.Data, r.cache = r.cache, r.Data` doesn't work?

File: query/raw_result.go (hunk 2 above)
Comment (xichen2020) [reply query/raw_result.go]: Yeah this has been handled in the segment query logic.

File: query/raw_result.go (hunk 3 above)
Comment (xichen2020) [reply query/raw_result.go]: Good point, will change.

**Pull request merged**